### ISSUE TITLE
[geonetwork] Missing HTTP parameter ES_HOST to fill the template

### ISSUE THREAD

**Issue by jeanpommier** [georchestra/georchestra#3960]
When starting a brand new georchestra instance, using the docker-compose. I'm seeing in the logs the following error:
`javax.servlet.ServletException: Missing HTTP parameter ES_HOST to fill the template`.

**Comment by jeanpommier** [comment]
I believe this is due to the fact that when you add a WMS layer in a metadata record, currently, [GN tries a WFS request](https://github.com/geonetwork/core-geonetwork/issues/4168) to fetch more information if available. This is apparently when the error appears on the logs

**Comment by jeanpommier** [comment]
And, of course, the core of the pb is env. variable subsitution, that doesn't seem to be done for this parameter

**Comment by jeanpommier** [comment]
This is https://github.com/georchestra/datadir/blob/docker-master/geonetwork/geonetwork.properties#L97 not being resolved. Can't figure out why yet

**Comment by landryb** [comment]
> I believe this is due to the fact that when you add a WMS layer in a metadata record, currently, [GN tries a WFS request](https://github.com/geonetwork/core-geonetwork/issues/4168) to fetch more information if available. This is apparently when the error appears on the logs

the corresponding GN issue talks about adding a layer to a map, but i can confirm i'm seeing a bunch of POST requests on `/geonetwork/index/features?_=_search` when editing a MD that contains a link to a WMS layer. This is especially confusing as for some reason it triggers an authentication popup which confuses all users ...

maybe for the docker variant it's an issue about ES_HOST not being templated for some reason, but the underlying geonetwork issue also affects non-docker deployments ?

**Comment by jeanmi151** [comment]
@jeanpommier we change a lot of things in the docker compose, i believe you run old config : 
https://github.com/georchestra/docker/blob/master/.envs-hosts#L14

**Comment by landryb** [comment]
i've checked and on all my instances, the `gn-features` ES index does exist but seems empty. At least it's not a case of trying to hit a non-existent index..

**Comment by jeanpommier** [comment]
> i've checked and on all my instances, the `gn-features` ES index does exist but seems empty. At least it's not a case of trying to hit a non-existent index..

Thanks @landryb for the check. But I'm 100% certaint this is a docker/k8s -only issue. Environment variable substitution that doesn't happen on the parameter mentioned just above

**Comment by jeanpommier** [comment]
If I replace the `${ES_HOST}` part by the actual value, the error message disappears

**Comment by pmauduit** [comment]
> @pmauduit do you know if environment variables are supposed to be resolved there ? https://github.com/georchestra/geonetwork/blame/83a493242fb8bde91d9e45c9e4e5e56a4be258c1/web/src/main/java/org/fao/geonet/proxy/GeorchestraPropertyResolver.java#L29-L34

no, it look like "pure java properties" parsing code, so it does not come with all the spring logic of recursively resolve variables, e.g.:

```groovy
import java.nio.file.Paths
File props = Paths.get("/tmp", "dudu.properties").toFile()

// Prints the file content
println props.text

// Parses the file as Properties
Properties gnProps = new Properties()
gnProps.load(props.newInputStream())
println gnProps
```

=>

```
flup=${USER}

[flup:${USER}]
```

**Comment by jeanpommier** [comment]
Thanks, that's what I thought. Do you know of an easy way to include the Spring logic there ? This would probably be the best fix

**Comment by landryb** [comment]
@jeanpommier in the end my issue is different, i've setupped authentication on ES, and the featureproxy doesnt support basic-auth so doesnt reuses `es.username` and `es.password`.

**Comment by jeanpommier** [comment]
Ow, sounds like bad luck. Maybe just wait for https://github.com/geonetwork/core-geonetwork/issues/4168 to be solved. I'd say it's mostly log pollution. I believe the request is mostly used for the internal map viewer to possibly index the features and allow to search in the data itself. Which AFAIK we are not using anywhere in geOrchestra tools.

**Comment by jeanpommier** [comment]
To get back to this ticket, unless we figure out a way to apply the Spring logic in the file, I can see 2 possible hacks:

- I can propose a very dirty fix, by inserting 

```
        String prop = gnProps.getProperty(name);
        if (name.equals("es.featureproxy.targeturi")) {
            prop = prop.replaceAll(Pattern.quote("${ES_HOST}"), System.getenv("ES_HOST"));
        }
        return prop;
```
in https://github.com/georchestra/geonetwork/blame/83a493242fb8bde91d9e45c9e4e5e56a4be258c1/web/src/main/java/org/fao/geonet/proxy/GeorchestraPropertyResolver.java

- Or we rewrite the geonetwork.properties file differently (don't use any java var in it but only  env vars) and apply the envsubst hack on the file. This should work
